tests/math/test_trace_distance_math.py

```diff
@@ -0,0 +1,288 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unit tests for differentiable trace distance.
+"""
+
+import numpy as onp
+import pytest
+
+import pennylane as qml
+from pennylane import numpy as np
+
+pytestmark = pytest.mark.all_interfaces
+
+tf = pytest.importorskip("tensorflow", minversion="2.1")
+torch = pytest.importorskip("torch")
+jax = pytest.importorskip("jax")
+jnp = pytest.importorskip("jax.numpy")
+
+
+class TestTraceDistanceMath:
+    """Tests for the trace distance function between two states (state vectors or density matrix)."""
+
+    state0_state1_td = [
+        # Mat-Mat-TD
+        (([[1, 0], [0, 0]], [[0.5, 0], [0, 0.5]]), 0.5),
+        (([[0, 0], [0, 1]], [[0.5, 0], [0, 0.5]]), 0.5),
+        (([[1, 0], [0, 0]], [[0.5, 0.5], [0.5, 0.5]]), np.sqrt(2) / 2),
+        (([[0, 0], [0, 1]], [[0.5, 0.5], [0.5, 0.5]]), np.sqrt(2) / 2),
+        (
+            (
+                [[0.5, 0, 0, 0.5], [0, 0, 0, 0], [0, 0, 0, 0], [0.5, 0, 0, 0.5]],
+                [
+                    [0.25, 0.25, 0.25, 0.25],
+                    [0.25, 0.25, 0.25, 0.25],
+                    [0.25, 0.25, 0.25, 0.25],
+                    [0.25, 0.25, 0.25, 0.25],
+                ],
+            ),
+            np.sqrt(2) / 2,
+        ),
+        # Batch-Mat-TD
```

Comment:
Can we separate the batched examples into a separate test from the non-batched examples, e.g. `test_broadcast_td`? Just so that if a test fails it's immediately clear what component isn't working. Also this list is massive and it would more readable to break it apart.